dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/MultipartRequestBody.java

```diff
@@ -0,0 +1,153 @@
+        super(entity);
+    }
+
+    public static final class Builder {
+
+        private final MultipartEntityBuilder builder = MultipartEntityBuilder.create();
+
+        private Builder() {}
+
+        @VisibleForTesting
+        Builder boundary(String boundary) {
+            builder.setBoundary(boundary);
+            return this;
+        }
+
+        public Builder addPart(Part part) {
+            builder.addPart(part.part);
+            return this;
+        }
+
+        public MultipartRequestBody build() {
+            return new MultipartRequestBody(builder.build());
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static ContentBodyPartBuilder contentBodyPartBuilder(ContentBody contentBody) {
+        return new ContentBodyPartBuilder(contentBody);
+    }
+
+    public static FormBodyPartBuilder formBodyPartBuilder(String name, ContentBody contentBody) {
+        return new FormBodyPartBuilder(name, contentBody);
+    }
+
+    public static final class FormBodyPartBuilder {
+        private final ContentBodyAdapter bodyAdapter;
+        private final org.apache.hc.client5.http.entity.mime.FormBodyPartBuilder builder;
+
+        private FormBodyPartBuilder(String name, ContentBody unsafeContentBody) {
+            bodyAdapter = new ContentBodyAdapter(unsafeContentBody);
+            builder = org.apache.hc.client5.http.entity.mime.FormBodyPartBuilder.create(name, bodyAdapter);
+        }
+
+        public FormBodyPartBuilder fileName(String fileName) {
+            bodyAdapter.setFileName(fileName);
+            return this;
+        }
+
+        public Part build() {
+            return new Part(builder.build());
+        }
+    }
+
+    public static final class ContentBodyPartBuilder {
+        private final MultipartPartBuilder builder;
+
+        private ContentBodyPartBuilder(ContentBody unsafeContentBody) {
+            this.builder = MultipartPartBuilder.create(new ContentBodyAdapter(unsafeContentBody));
+        }
+
+        public ContentBodyPartBuilder addHeaderValue(String key, String value) {
+            builder.addHeader(key, value);
+            return this;
+        }
+
+        public Part build() {
+            return new Part(builder.build());
+        }
+    }
+
+    public static final class Part {
+        private final MultipartPart part;
+
+        private Part(MultipartPart part) {
+            this.part = part;
+        }
+    }
+
+    private static final class ContentBodyAdapter extends AbstractContentBody {
+
+        private final ContentBody unsafeRequestBody;
+
+        @Nullable
+        private String fileName;
+
+        private ContentBodyAdapter(ContentBody requestBody) {
+            super(Preconditions.checkNotNull(
+                    ContentType.parse(requestBody.contentType()),
+                    "Invalid content type",
+                    UnsafeArg.of("contentType", requestBody.contentType())));
```

Comment:
content-type should be safe, even if invalid.

Reply:
Done